#실습 과제 4

# 밑에 코드에서 주어진 데이터를 사용하여 문제 1과 문제 2를 푸시오.

In [1]:
#모든 ramdom state는 42로 고정하시오.

from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

data = load_wine()
X_train, X_test, y_train, y_test = train_test_split(data.data, data.target, test_size= 0.2, random_state = 42)
X_hold, X_test, y_hold, y_test = train_test_split(X_test, y_test, test_size= 0.5, random_state = 11)

# 문제 1
# 2가지 LogisticRegression classifier를 만들고 해당 모델들의 train dataset에 대한 accuracy와 holdout validation set의 accuracy를 비교하시오.

### 첫번째 모델은 regularization(penalty)를 적용하지 않은 모델
### 두번째 모델은 L2 regularization(penalty)를 적용한 모델

In [2]:
# 필요한 라이브러리와 데이터를 불러옵니다.
from sklearn.datasets import load_wine
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

data = load_wine()
# 데이터를 train set과 test set으로 나눕니다.
X_train, X_test, y_train, y_test = train_test_split(data.data, data.target, test_size= 0.2, random_state = 42)
# test set을 다시 holdout validation set과 test set으로 나눕니다.
X_hold, X_test, y_hold, y_test = train_test_split(X_test, y_test, test_size= 0.5, random_state = 11)

# 두 개의 LogisticRegression 모델을 생성합니다.
model1 = LogisticRegression(random_state=42)
model2 = LogisticRegression(penalty='l2', random_state=42)

# 두 모델을 training dataset으로 학습시킵니다.
model1.fit(X_train, y_train)
model2.fit(X_train, y_train)

# training set과 holdout validation set에 대한 정확도를 평가합니다.
train_acc1 = accuracy_score(y_train, model1.predict(X_train))
train_acc2 = accuracy_score(y_train, model2.predict(X_train))
holdout_acc1 = accuracy_score(y_hold, model1.predict(X_hold))
holdout_acc2 = accuracy_score(y_hold, model2.predict(X_hold))

# 결과를 출력합니다.
print("Model 1 (no regularization) - Training accuracy: {:.2f}, Holdout validation accuracy: {:.2f}".format(train_acc1, holdout_acc1))
print("Model 2 (L2 regularization) - Training accuracy: {:.2f}, Holdout validation accuracy: {:.2f}".format(train_acc2, holdout_acc2))

Model 1 (no regularization) - Training accuracy: 0.96, Holdout validation accuracy: 1.00
Model 2 (L2 regularization) - Training accuracy: 0.96, Holdout validation accuracy: 1.00


C:\Users\ADMIN\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
C:\Users\ADMIN\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.10_qbz5n2kfra8p0\LocalCache\local-packages\Python310\site-packages\sklearn\linear_model\_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stab

# 문제 2
## 주어진 파라미터를 이용하여 SVM classifier를 학습시키고 train dataset에 대한 accuracy와 holdout validation에 대한 accuracy를 출력하시오.
## 해당 결과에 나타나는 현상에 대해서 서술하고 그러한 현상을 극복하기 위한 방법에 대해 간략히 서술하시오.

In [3]:
from sklearn.svm import SVC

gamma = 5
C = 1000

# 주어진 파라미터를 이용하여 SVM 모델을 생성합니다.
model = SVC(kernel='rbf', gamma=gamma, C=C, random_state=42)

# 학습 데이터셋으로 모델을 학습시킵니다.
model.fit(X_train, y_train)

# 학습 데이터셋과 holdout validation set에 대한 정확도를 평가합니다.
train_acc = accuracy_score(y_train, model.predict(X_train))
holdout_acc = accuracy_score(y_hold, model.predict(X_hold))

# 결과를 출력합니다.
print("SVM classifier - Training accuracy: {:.2f}, Holdout validation accuracy: {:.2f}".format(train_acc, holdout_acc))


SVM classifier - Training accuracy: 1.00, Holdout validation accuracy: 0.28


# 문제 3
## 문제 1번과 2번에서 학습한 모델들의 각 holdout validation set의 accuracy를 비교하고 어떤 모델을 final 모델로 사용하는 것이 적절한 것인지 설명하시오.


첫 번째 모델과 두 번째 모델은 모두 훈련 정확도가 0.96이고, 홀드아웃 검증 정확도가 1.00으로 매우 높은 성능을 보이고 있습니다. 이는 모델이 훈련 데이터에 잘 맞춰져 있으며, 검증 데이터에 대해서도 높은 정확도를 보이고 있음을 의미합니다. 그러나 첫 번째 모델은 정규화를 적용하지 않았기 때문에, 새로운 데이터에 대한 일반화 성능이 떨어질 수 있습니다. 반면에 두 번째 모델은 L2 정규화를 적용하여 모델의 복잡도를 제한함으로써 새로운 데이터에 대한 일반화 성능이 더 좋을 것으로 예상됩니다.

SVM 분류기의 경우, 훈련 정확도가 1.00으로 매우 높지만, 홀드아웃 검증 정확도가 0.28로 매우 낮습니다. 이는 모델이 훈련 데이터에 과적합되어 있으며, 새로운 데이터에 대한 일반화 성능이 매우 떨어진다는 것을 의미합니다.

따라서 세 가지 모델 중에서 final 모델로 선택하는 것이 적절한 것은 두 번째 모델, 즉 L2 정규화를 적용한 모델입니다. 이 모델은 훈련 데이터에 잘 맞춰져 있을 뿐만 아니라, 홀드아웃 검증 데이터에 대해서도 높은 정확도를 보이고 있으며, 정규화를 통해 모델의 복잡도를 제한함으로써 새로운 데이터에 대한 일반화 성능도 좋을 것으로 예상됩니다.